common/admin/manager/module_test.go

```diff
@@ -0,0 +1,2388 @@
+package manager
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ceph/go-ceph/internal/admintest"
+	"github.com/ceph/go-ceph/internal/commands"
+)
+
+var radosConnector = admintest.NewConnector()
+
+func TestParseModuleInfo(t *testing.T) {
+	t.Run("valid", func(t *testing.T) {
+		r := commands.NewResponse([]byte(cephMgrModuleLsJSON1), "", nil)
+		mi, err := parseModuleInfo(r)
+		assert.NoError(t, err)
+		if assert.Len(t, mi.EnabledModules, 7) {
+			assert.Contains(t, mi.EnabledModules, "mirroring")
+			assert.Contains(t, mi.EnabledModules, "nfs")
+			assert.NotContains(t, mi.EnabledModules, "progress")
+		}
+		if assert.Len(t, mi.AlwaysOnModules, 10) {
+			assert.Contains(t, mi.AlwaysOnModules, "balancer")
+			assert.Contains(t, mi.AlwaysOnModules, "volumes")
+			assert.NotContains(t, mi.AlwaysOnModules, "notebook")
+		}
+		if assert.Len(t, mi.DisabledModules, 19) {
+			assert.Equal(t, mi.DisabledModules[0].Name, "alerts")
+			assert.True(t, mi.DisabledModules[0].CanRun)
+
+			assert.Contains(t, mi.DisabledModules[4].Name, "influx")
+			assert.False(t, mi.DisabledModules[4].CanRun)
+
+			assert.Contains(t, mi.DisabledModules[10].Name, "osd_support")
+			assert.True(t, mi.DisabledModules[10].CanRun)
+		}
+	})
+	t.Run("error", func(t *testing.T) {
+		r := commands.NewResponse(nil, "", errors.New("foo"))
+		mi, err := parseModuleInfo(r)
+		assert.Error(t, err)
+		assert.Nil(t, mi)
+	})
+}
+
+func TestEnableDisableModule(t *testing.T) {
+	ra := radosConnector.Get(t)
+	mgrAdmin := NewFromConn(ra)
+
+	mi, err := mgrAdmin.ListModules()
+	require.NoError(t, err)
+
+	var toEnable string
+dmloop:
+	for _, entry := range mi.DisabledModules {
+		if !entry.CanRun {
+			continue
+		}
+		switch entry.Name {
+		// not all modules in the disabled list work (in our container?).
+		// here's a short list our test is allowed to try enabling
+		case "hello", "mirroring", "rgw", "prometheus", "dashboard":
+			toEnable = entry.Name
+			break dmloop
+		default:
+		}
+	}
+	require.NotEqual(t, "", toEnable)
+
+	err = mgrAdmin.EnableModule(toEnable, false)
+	require.NoError(t, err)
+	waitFor(t, mgrAdmin, toEnable)
+
+	// put it back to disabled state
+	err = mgrAdmin.DisableModule(toEnable)
+	require.NoError(t, err)
+}
+
+func waitFor(t *testing.T, mgrAdmin *MgrAdmin, expect string) {
+	for i := 0; i < 30; i++ {
+		modinfo, err := mgrAdmin.ListModules()
+		require.NoError(t, err)
+		for _, emod := range modinfo.EnabledModules {
+			if emod == expect {
+				return
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	t.Fatalf("timed out waiting for %s module", expect)
+}
+
+var cephMgrModuleLsJSON1 = `
```

Comment:
I would expect this to be a `const`

Reply:
But it is for testing only, so it does not matter too much.

Reply:
thanks. I will try to keep that in mind for the next time